infrastructure/src/main/java/org/wso2/testgrid/infrastructure/providers/aws/KibanaDashboardBuilder.java

```diff
@@ -0,0 +1,189 @@
+            kibanaDashboardBuilder = new KibanaDashboardBuilder();
+        }
+        return kibanaDashboardBuilder;
+    }
+
+    /**
+     * This method is responsible for building the Kibana dashboard for the available stacks.
+     *
+     * @param currentInstancesMap Map of instanceName and instance-id
+     * @param currentStackName Cloudformation stack name
+     * @param shortenURL weather to shorten URL or not, primaryliy used for unit testing purposes
+     * @return Optional of dashboard link shortened to make it compatible with database column
+     * restrictions
+     */
+    public Optional<String> buildDashBoard(Map<String, String> currentInstancesMap
+            , String currentStackName, boolean shortenURL) {
+
+        if (kibanaEndpoint == null || dashboardCtxFormat == null || instanceLogFilterFormat == null) {
+            logger.warn("Kibana endpoint configuration not found in testgrid config. Server log view may not work!" +
+                    "Kibana Endpoint : " + kibanaEndpoint +
+            "\nDashbboardCtxFormat : " + dashboardCtxFormat +
+            "\ninstanceLogFilterFormat : " + instanceLogFilterFormat);
+            return Optional.empty();
+        }
+        //TODO implement filtering unwanted nodes with no log input i.e BastianNode , PuppetMaster
+        allInstances.put(currentStackName, currentInstancesMap);
+        String instanceLogFilter;
+        StringJoiner filtersStr = new StringJoiner(",");
+
+        for (Map.Entry<String, Map<String, String>> allInstancesEntry : allInstances.entrySet()) {
+            Map<String, String> instanceMap = allInstancesEntry.getValue();
+            for (Map.Entry<String, String> entry : instanceMap.entrySet()) {
+                instanceLogFilter = instanceLogFilterFormat
+                        .replaceAll("#_INSTANCE_ID_#", entry.getKey())
+                        .replaceAll("#_LABEL_#", entry.getValue())
+                        .replaceAll("#_STACK_NAME_#", allInstancesEntry.getKey());
+                filtersStr.add(instanceLogFilter);
+            }
+        }
+
+        StringJoiner allLogsStr = new StringJoiner(",");
+        StringJoiner allLogsJson = new StringJoiner(",");
+
+        for (String stackName : allInstances.keySet()) {
+            allLogsStr.add(allLogsFilterEncodedSection.replaceAll("#_STACK_NAME_#", stackName));
+            allLogsJson.add(allLogsFilterJsonSection.replaceAll("#_STACK_NAME_#", stackName));
+        }
+        allLogsFilter = allLogsFilter
+                .replaceAll("#_ALL_LOGS_FILTER_SECTION_#", allLogsStr.toString())
+                .replaceAll("#_REPEATABLE_ALL_LOGS_JSON_SECTION_#", allLogsJson.toString());
+        filtersStr.add(allLogsFilter);
+        String logDownloadCtx = dashboardCtxFormat.replace("#_NODE_FILTERS_#", filtersStr.toString())
+                .replaceAll("#_STACK_NAME_#", currentStackName);
+
+        if (shortenURL) {
+            return shortenKibanaURL(logDownloadCtx);
```

Comment:
Further, if shortening failed, then we should return the original url as it is. Shortening is not a critical requirement.

Reply:
if the shortening fails the original URL is being logged in the shortening method. 
The original URL is not returned back because it will break the TG DB persist due to character level exceeding the limit.